File: templates/Uwp/Features/ShareSource/Models/ShareSourceFeatureItem.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

using Windows.Storage;

namespace Param_RootNamespace.Models
{
    internal enum ShareSourceFeatureItemType
    {
        Text = 0,
        WebLink = 1,
        ApplicationLink = 2,
        Html = 3,
        Image = 4,
        StorageItems = 5,
        DeferredContent = 6
    }

    internal class ShareSourceFeatureItem
    {
        public ShareSourceFeatureItemType DataType { get; }

        public string Text { get; private set; }

        public Uri WebLink { get; private set; }

        public Uri ApplicationLink { get; private set; }

        public string Html { get; private set; }

        public StorageFile Image { get; private set; }

        public IEnumerable<IStorageItem> StorageItems { get; private set; }

        public string DeferredDataFormatId { get; private set; }

        public Func<Task<object>> GetDeferredDataAsyncFunc { get; private set; }

        private ShareSourceFeatureItem(ShareSourceFeatureItemType dataType)
        {
            DataType = dataType;
        }

        internal static ShareSourceFeatureItem FromText(string text)
        {
            return new ShareSourceFeatureItem(ShareSourceFeatureItemType.Text)
            {
                Text = text
            };
        }

        internal static ShareSourceFeatureItem FromWebLink(Uri webLink)
        {
            return new ShareSourceFeatureItem(ShareSourceFeatureItemType.WebLink)
            {
                WebLink = webLink
            };
        }

        internal static ShareSourceFeatureItem FromApplicationLink(Uri applicationLink)
        {
            return new ShareSourceFeatureItem(ShareSourceFeatureItemType.ApplicationLink)
            {
                ApplicationLink = applicationLink
            };
        }

        internal static ShareSourceFeatureItem FromHtml(string html)
        {
            return new ShareSourceFeatureItem(ShareSourceFeatureItemType.Html)
            {
                Html = html
            };
        }

        internal static ShareSourceFeatureItem FromImage(StorageFile image)
        {
            return new ShareSourceFeatureItem(ShareSourceFeatureItemType.Image)
            {
                Image = image
            };
        }

        internal static ShareSourceFeatureItem FromStorageItems(IEnumerable<IStorageItem> storageItems)
        {
            return new ShareSourceFeatureItem(ShareSourceFeatureItemType.StorageItems)
            {
                StorageItems = storageItems
            };
        }

        internal static ShareSourceFeatureItem FromDeferredContent(string deferredDataFormatId, Func<Task<object>> getDeferredDataAsyncFunc)
        {
            return new ShareSourceFeatureItem(ShareSourceFeatureItemType.DeferredContent)
            {
                DeferredDataFormatId = deferredDataFormatId,
                GetDeferredDataAsyncFunc = getDeferredDataAsyncFunc
            };
        }
    }
}

File: code/TemplateStudioForUWP/Templates/Uwp/Features/ShareSource._VB/Models/ShareSourceFeatureItem.vb
Imports Windows.Storage

Namespace Models

    Friend Enum ShareSourceFeatureItemType
        Text = 0
        WebLink = 1
        ApplicationLink = 2
        Html = 3
        Image = 4
        StorageItems = 5
        DeferredContent = 6
    End Enum

    Friend Class ShareSourceFeatureItem

        Public Property DataType As ShareSourceFeatureItemType

        Public Property Text As String

        Public Property WebLink As Uri

        Public Property ApplicationLink As Uri

        Public Property Html As String

        Public Property Image As StorageFile

        Public Property StorageItems As IEnumerable(Of IStorageItem)

        Public Property DeferredDataFormatId As String

        Public Property GetDeferredDataAsyncFunc As Func(Of Task(Of Object))

        Private Sub New(dataType As ShareSourceFeatureItemType)
            DataType = dataType
        End Sub

        Friend Shared Function FromText(text As String) As ShareSourceFeatureItem
            Return New ShareSourceFeatureItem(ShareSourceFeatureItemType.Text) With {.Text = text}
        End Function

        Friend Shared Function FromWebLink(webLink As Uri) As ShareSourceFeatureItem
            Return New ShareSourceFeatureItem(ShareSourceFeatureItemType.WebLink) With {.WebLink = webLink}
        End Function

        Friend Shared Function FromApplicationLink(applicationLink As Uri) As ShareSourceFeatureItem
            Return New ShareSourceFeatureItem(ShareSourceFeatureItemType.ApplicationLink) With {.ApplicationLink = applicationLink}
        End Function

        Friend Shared Function FromHtml(html As String) As ShareSourceFeatureItem
            Return New ShareSourceFeatureItem(ShareSourceFeatureItemType.Html) With {.Html = html}
        End Function

        Friend Shared Function FromImage(image As StorageFile) As ShareSourceFeatureItem
            Return New ShareSourceFeatureItem(ShareSourceFeatureItemType.Image) With {.Image = image}
        End Function

        Friend Shared Function FromStorageItems(storageItems As IEnumerable(Of IStorageItem)) As ShareSourceFeatureItem
            Return New ShareSourceFeatureItem(ShareSourceFeatureItemType.StorageItems) With {.StorageItems = storageItems}
        End Function

        Friend Shared Function FromDeferredContent(deferredDataFormatId As String, getDeferredDataAsyncFunc As Func(Of Task(Of Object))) As ShareSourceFeatureItem
            Return New ShareSourceFeatureItem(ShareSourceFeatureItemType.DeferredContent) With {.DeferredDataFormatId = deferredDataFormatId, .GetDeferredDataAsyncFunc = getDeferredDataAsyncFunc}
        End Function
    End Class
End Namespace
